collector/infiniband_linux.go

```diff
@@ -0,0 +1,167 @@
+}
+
+func NewInfiniBandCollector() (Collector, error) {
+	var i infinibandCollector
+
+	// Filenames of all InfiniBand counter metrics including a detailed description.
+	i.counters = map[string]string{
+		"link_downed":            "Number of times the link failed to recover from an error state and went down",
+		"link_error_recovery":    "Number of times the link successfully recovered from an error state",
+		"multicast_rcv_packets":  "Number of multicast packets received (including errors)",
+		"multicast_xmit_packets": "Number of multicast packets transmitted (including errors)",
+		"port_rcv_data":          "Number of data octets received on all links",
+		"port_xmit_data":         "Number of data octets transmitted on all links",
+		"unicast_rcv_packets":    "Number of unicast packets received (including errors)",
+		"unicast_xmit_packets":   "Number of unicast packets transmitted (including errors)",
+	}
+
+	subsystem := "infiniband"
+	i.metricDescs = make(map[string]*prometheus.Desc)
+
+	for metric, description := range i.counters {
+		i.metricDescs[metric] = prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, metric+"_total"),
+			description,
+			[]string{"device", "port"},
+			nil)
+	}
+
+	return &i, nil
+}
+
+// Retrieve a list of InfiniBand devices.
+func infinibandDevices(infinibandPath string) ([]string, error) {
+	devices, err := filepath.Glob(filepath.Join(infinibandPath, "/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(devices) < 1 {
+		log.Debugf("Unable to detect InfiniBand devices")
+		err = infinibandNoDevicesFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equate to the device names.
+	for i, device := range devices {
+		devices[i] = filepath.Base(device)
+	}
+
+	return devices, nil
+}
+
+// Retrieve a list of ports for the InfiniBand device.
+func infinibandPorts(infinibandPath, device string) ([]string, error) {
+	ports, err := filepath.Glob(filepath.Join(infinibandPath, device, "ports/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ports) < 1 {
+		log.Debugf("Unable to detect ports for %s", device)
+		err = infinibandNoPortsFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equates to the port numbers.
+	for i, port := range ports {
+		ports[i] = filepath.Base(port)
+	}
+
+	return ports, nil
+}
+
+func readMetric(directory, metricFile string) (uint64, error) {
+	metric, err := readUintFromFile(filepath.Join(directory, metricFile))
+
+	if err != nil {
+		log.Debugf("Error reading " + metricFile + " file")
+		return 0, err
+	}
+
+	return metric, nil
+}
+
+func (c *infinibandCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	devices, err := infinibandDevices(sysFilePath(infinibandPath))
+
+	// If no devices are found or another error is raised while attempting to find devices,
+	// InfiniBand is likely not installed and the collector should be skipped.
+	if err == infinibandNoDevicesFound || err != nil {
```

Comment:
In the case that this error isn't "no devices found", you'll want to return it.  A `switch` statement is nice.  The empty `case nil:` results in an implicit break.  Go switches only "fall through" with the `fallthrough` keyword.

```go
switch err {
case nil:
case infinibandNoDevicesFound:
    return nil
default:
    return err
}
```